Panic in sys_dup3 due to arithmetic overflow in SlotVec::put_at when new_fd is usize::MAX

### Describe the bug

The kernel panics with an "attempt to add with overflow" error when the dup3 syscall is invoked with new_fd set to usize::MAX

# https://github.com/asterinas/asterinas/blob/77f0b0e746ce2df0066b007fb050fc5be0e86e47/kernel/libs/aster-util/src/slot_vec.rs#L84

### To Reproduce

1. Compile the program and run.

```
#include <unistd.h>
#include <fcntl.h>
#include <sys/syscall.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>

int main() {
    // Open a file to get a valid file descriptor
    int fd = open("/dev/null", O_RDONLY);
    if (fd < 0) {
        perror("open failed");
        return 1;
    }

    // Attempt to duplicate the file descriptor to a very large number
    // This will trigger the potential arithmetic overflow in the kernel's
    // dup function when new_fd is close to usize::MAX
    // The overflow occurs at: for idx in new_fd + 1..self.len()
    long new_fd = 0xFFFFFFFFFFFFFFFFUL; // Maximum value for 64-bit unsigned
    
    // Use dup3 syscall which allows specifying the target fd
    // This maps to the dup function in the kernel where the overflow occurs
    int result = syscall(SYS_dup3, fd, new_fd, 0);
    
    if (result == -1) {
        perror("dup3 failed");
        // Expected to fail, but the kernel should panic due to arithmetic overflow
    } else {
        printf("dup3 succeeded, fd: %d\n", result);
    }

    close(fd);
    return 0;
}
```

### Environment

- Official docker asterinas/asterinas:0.17.0-20251217
- Asterinas version: main [Fix panic in `add_posix_timer` · asterinas/asterinas@d9f4229](https://github.com/asterinas/asterinas/commit/d9f4229e6c690c48f0a9b055665aadaefcf3d702)

### Logs

```
~ # ./ex690___home__yuchen__workspace__asterinas__kernel__src__fs__file_table.rs_51_24 
[    18.638] ERROR: Uncaught panic:
        attempt to add with overflow
        at /root/asterinas/kernel/libs/aster-util/src/slot_vec.rs:84
        on CPU 0 by thread Some(Thread { task: (Weak), data: Any { .. }, is_exited: false, cpu_affinity: AtomicIdSet { bits: [1], phantom: PhantomData<ostd::cpu::id::CpuId> }, sched_attr: SchedAttr { policy: SchedPolicyState { kind: AtomicSchedPolicyKind(2), policy: UnsafeCell { .. } }, last_cpu: AtomicCpuId(0), real_time: RealTimeAttr { prio: 99, time_slice: 31491940 }, fair: FairAttr { weight: 1024, pending_weight: 0, vruntime: 325372982 } } })
Printing stack trace:
   1: fn 0xffffffff88950e10 - pc 0xffffffff88950e5b / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd707e0;

   2: fn 0xffffffff8804c720 - pc 0xffffffff8804cf1d / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd70800;

   3: fn 0xffffffff8804e9e0 - pc 0xffffffff8804e9ea / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd70bd0;

   4: fn 0xffffffff8804b000 - pc 0xffffffff8804b00a / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd70be0;

   5: fn 0xffffffff88ae10a0 - pc 0xffffffff88ae10e6 / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd70bf0;

   6: fn 0xffffffff88aeb3c0 - pc 0xffffffff88aeb3e5 / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd70c30;

   7: fn 0xffffffff886acfc0 - pc 0xffffffff886ad07e / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd70c40;

   8: fn 0xffffffff88271c00 - pc 0xffffffff88271dde / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd70ce0;

   9: fn 0xffffffff88541850 - pc 0xffffffff88541986 / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd70ee0;

  10: fn 0xffffffff885421e0 - pc 0xffffffff885423df / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd70ff0;

  11: fn 0xffffffff883684f0 - pc 0xffffffff8839bd3f / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd710f0;

  12: fn 0xffffffff88365940 - pc 0xffffffff883659da / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd844f0;

  13: fn 0xffffffff88302b40 - pc 0xffffffff883033f1 / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd84650;

  14: fn 0xffffffff883035b0 - pc 0xffffffff8830360e / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd84b40;

  15: fn 0xffffffff88692b10 - pc 0xffffffff88692b3e / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd84cf0;

  16: fn 0xffffffff88692a40 - pc 0xffffffff88692a86 / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd84d20;

  17: fn 0xffffffff8850df50 - pc 0xffffffff8850df7c / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd84d90;

  18: fn 0xffffffff885903d0 - pc 0xffffffff8859040f / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd84dd0;

  19: fn 0xffffffff88303570 - pc 0xffffffff88303597 / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd84f40;

  20: fn 0xffffffff880c2c00 - pc 0xffffffff880c2c1d / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd84f80;

  21: fn 0xffffffff88963e10 - pc 0xffffffff88963e35 / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd84fa0;

  22: fn 0xffffffff889110c0 - pc 0xffffffff88911144 / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd84fd0;

  23: fn 0xffffffff889d94a9 - pc 0xffffffff889d94b2 / registers:

     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd85000;


     rax 0xffffdfffffd706f8; rdx 0xffffffff889899b2; rcx  0x100000201007574; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc7030; rbp                0x0; rsp 0xffffdfffffd85008;

OSDK] The kernel seems panicked. Parsing stack trace for source lines:
(  1) ostd::panic::print_stack_trace
          at /root/asterinas/ostd/src/panic.rs:113
(  2) aster_kernel::thread::oops::panic_handler
          at /root/asterinas/kernel/src/thread/oops.rs:130
(  3) __ostd_panic_handler
          at ??:?
(  4) __rustc::rust_begin_unwind
          at cawjknoo4v9jar9emdep6lx3w:?
(  5) core::panicking::panic_fmt
          at ??:?
(  6) core::panicking::panic_const::panic_const_add_overflow
          at ??:?
(  7) <aster_util::slot_vec::SlotVec<aster_kernel::fs::file_table::FileTableEntry>>::put_at
          at ??:?
(  8) <aster_kernel::fs::file_table::FileTable>::dup
          at /root/asterinas/kernel/src/fs/file_table.rs:61
(  9) aster_kernel::syscall::dup::do_dup3
          at /root/asterinas/kernel/src/syscall/dup.rs:72
( 10) aster_kernel::syscall::dup::sys_dup3
          at au0euf0e5mgxo7altn6e1e6l0:?
( 11) aster_kernel::syscall::arch::syscall_dispatch
          at /root/asterinas/kernel/src/syscall/mod.rs:217
( 12) aster_kernel::syscall::handle_syscall
          at /root/asterinas/kernel/src/syscall/mod.rs:374
( 13) aster_kernel::thread::task::create_new_user_task::{closure#0}
          at /root/asterinas/kernel/src/thread/task.rs:95
( 14) aster_kernel::thread::task::create_new_user_task::{closure#1}
          at /root/asterinas/kernel/src/thread/task.rs:126
( 15) unwinding::panicking::catch_unwind::do_call::<aster_kernel::thread::task::create_new_user_task::{closure#1}, ()>
          at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/unwinding-0.2.8/src/panicking.rs:56
( 16) unwinding::panicking::catch_unwind::<unwinding::panic::RustPanic, (), aster_kernel::thread::task::create_new_user_task::{closure#1}>
          at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/unwinding-0.2.8/src/panicking.rs:42
( 17) unwinding::panic::catch_unwind::<(), aster_kernel::thread::task::create_new_user_task::{closure#1}>
          at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/unwinding-0.2.8/src/panic.rs:87
( 18) aster_kernel::thread::oops::catch_panics_as_oops::<aster_kernel::thread::task::create_new_user_task::{closure#1}, ()>
          at /root/asterinas/kernel/src/thread/oops.rs:55
( 19) aster_kernel::thread::task::create_new_user_task::{closure#2}
          at /root/asterinas/kernel/src/thread/task.rs:131
( 20) <aster_kernel::thread::task::create_new_user_task::{closure#2} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
          at /root/.rustup/toolchains/nightly-2025-12-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250
( 21) <alloc::boxed::Box<dyn core::ops::function::FnOnce<(), Output = ()> + core::marker::Send> as core::ops::function::FnOnce<()>>::call_once
          at /root/.rustup/toolchains/nightly-2025-12-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2206
( 22) kernel_task_entry
          at /root/asterinas/ostd/src/task/mod.rs:196
( 23) core::panicking::assert_failed::<alloc::vec::Vec<core::ops::range::Range<usize>>, alloc::vec::Vec<core::ops::range::Range<usize>>>
          at /root/.rustup/toolchains/nightly-2025-12-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:384
make: *** [Makefile:309: run_kernel] Error 1
```

--- a/kernel/src/syscall/dup.rs
+++ b/kernel/src/syscall/dup.rs
@@ -56,12 +56,12 @@ fn do_dup3(
         return_errno!(Errno::EINVAL);
     }
 
-    if new_fd
+    if new_fd.cast_unsigned() as u64
         >= ctx
             .process
             .resource_limits()
             .get_rlimit(ResourceType::RLIMIT_NOFILE)
-            .get_cur() as FileDesc
+            .get_cur()
     {
         return_errno!(Errno::EBADF);
     }

